src/modules/0.0.3/aws_security_group/index.ts

```diff
@@ -180,53 +182,67 @@ export const AwsSecurityGroupModule: Module = new Module({
-            // AWS does not have a way to update the top-level SecurityGroup entity. You can
-            // update the various rules associated with it, but not the name or description of the
-            // SecurityGroup itself. This may seem counter-intuitive, but we only need to create
-            // the security group in AWS and *eventually* the old one will be removed. Why?
-            // Because on the second pass of the checking algorithm (it always performs another
-            // pass if it performed any change, and only stops once it determines nothing needs to
-            // be changed anymore), it will see a security group in AWS that it doesn't have a
-            // record for and then remove it since the database is supposed to be the source of
-            // truth. Further, because of the relations to the security group being by internal ID
-            // in the database instead of the string ID, anything depending on the old security
-            // group will be moved to the new one on the second pass. However, there is a unique
-            // constraint on the `GroupName`, so a temporary state with a random name may be
-            // necessary, so we try-catch this call and mutate as necessary.
-            try {
-              return await AwsSecurityGroupModule.mappers.securityGroup.cloud.create(e, ctx);
-            } catch (_) {
-              // We mutate the `GroupName` to something unique and unlikely to collide (we should
-              // be too slow to ever collide at a millisecond level). This path doesn't save back
-              // to the DB like create does (at least right now, if that changes, we need to
-              // rethink this logic here)
-              e.groupName = Date.now().toString();
-              return await AwsSecurityGroupModule.utils.sgCloudCreate([e], ctx, true);
+        update: async (es: SecurityGroup[], ctx: Context) => {
+          const out = [];
+          for (const e of es) {
+            // Special behavior here. You're not allowed to mess with the "default" SecurityGroup.
+            // You can mess with its rules, but not this record itself, so any attempt to update it
+            // is instead turned into *restoring* the value in the database to match the cloud value
+            if (e.groupName === 'default') {
+              // Because updates are based on the `groupId` matching but not some other property,
+              // we can be sure that the security group rules for the default security group are
+              // properly associated so we don't need to do anything about them here, just restore
+              // the other properties
+              const cloudRecord = ctx?.memo?.cloud?.SecurityGroup?.[e.groupId ?? ''];
+              cloudRecord.id = e.id;
+              await AwsSecurityGroupModule.mappers.securityGroup.db.update(cloudRecord, ctx);
+              ctx.memo.db.SecurityGroup[cloudRecord.groupId] = cloudRecord; // Force the cache
```

Comment:
I'm actually surprised that it somehow worked *without* this before.

The `apply` loop grabs basically the entire database and every cloud record it can get its hands on (right now, until we implement the "faster apply" task) and builds a diff of the two to figure out what changes it should make.

Once it is done applying those changes (like this one here), it drops *only the cloud cache* and re-builds it and does another check. Once it is satisfied that there are no differences between the two, it drops both caches and does another check in case there are discrepancies still with the actual database records.

We use that behavior here in the security groups module so you can change the description of a non-default security group without changing the name of the security group because AWS doesn't actually allow that in reality. We go through a fake security group name in one step that's stored only in the in-memory cache of the DB-side of things to get everything depending on that security group to migrate over to this temporary one, and then the cache is dropped and now there's no conflict on simply creating the "real" security group name again with the changed description.

But this path in here where we intentionally *block* changes to the default security group would not pick up on the restoration of the record because the cache still says "do the update" and that's what it was getting stuck on in an infinite loop.

I'm not sure *how* it was working before when we did operations in parallel, because it should have seen the "wrong" value in the cache and gotten into an infinite loop, too. Which just shows a big part of why we needed to get rid of these `Promise.all`s; they run in "parallel" (closer to reality than you'd expect since they're all operating on external APIs) and there's weird data state race conditions that I can't reason about, and this change eliminates it: the log order is the same as the apply order, and couple that with the apply algo we have I can walk through what each diff will do by hand if necessary. :)

Reply:
hmm maybe with the parallel execution the cache was getting correctly updated for another path? definitely unexpected.